pkg/dns/ibm/dns.go

```diff
@@ -0,0 +1,194 @@
+)
+
+var (
+	_   dns.Provider = &Provider{}
+	log              = logf.Logger.WithName("dns")
+)
+
+type Provider struct {
+	dnsServices map[string]dnsclient.DnsClient
+}
+
+// Config is the necessary input to configure the manager.
+type Config struct {
+	APIKey    string
+	CISCRN    string
+	UserAgent string
+	Zones     []string
+}
+
+func NewProvider(config Config) (*Provider, error) {
+	if len(config.Zones) < 1 {
+		return nil, fmt.Errorf("missing zone data")
+	}
+	authenticator := &core.IamAuthenticator{
+		ApiKey: config.APIKey,
+	}
+	provider := &Provider{}
+
+	for _, zone := range config.Zones {
+		options := &dnsrecordsv1.DnsRecordsV1Options{
+			Authenticator:  authenticator,
+			URL:            "https://api.private.cis.cloud.ibm.com/v1",
+			Crn:            &config.CISCRN,
+			ZoneIdentifier: &zone,
+		}
+
+		dnsService, err := dnsrecordsv1.NewDnsRecordsV1(options)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create a new DNS Service instance: %v", err)
+		}
+		dnsService.EnableRetries(3, 5*time.Second)
+		dnsService.Service.SetUserAgent(config.UserAgent)
+
+		provider.dnsServices[zone] = dnsService
+	}
+
+	if err := validateDNSServices(provider); err != nil {
+		return nil, fmt.Errorf("failed to validate ibm dns services: %v", err)
+	}
+	return provider, nil
+}
+
+// validateDNSServices validates that provider clients can communicate with
+// associated API endpoints by having each client make a get DNS records call.
+func validateDNSServices(provider *Provider) error {
+	var errs []error
+	maxItems := int64(1)
+	for _, dnsService := range provider.dnsServices {
+		opt := dnsService.NewListAllDnsRecordsOptions()
+		opt.PerPage = &maxItems
+		if _, _, err := dnsService.ListAllDnsRecords(opt); err != nil {
+			errs = append(errs, fmt.Errorf("failed to get dns records: %v", err))
+		}
+	}
+	return kerrors.NewAggregate(errs)
+}
+
+func (p *Provider) Ensure(record *iov1.DNSRecord, zone configv1.DNSZone) error {
+	return p.createOrUpdateDNSRecord(record, zone)
+}
+
+func (p *Provider) Replace(record *iov1.DNSRecord, zone configv1.DNSZone) error {
+	return p.createOrUpdateDNSRecord(record, zone)
+}
+
+func (p *Provider) Delete(record *iov1.DNSRecord, zone configv1.DNSZone) error {
+	if err := validateInputDNSData(record, zone); err != nil {
+		return fmt.Errorf("delete: invalid dns input data: %v", err)
+	}
+	dnsService := p.dnsServices[zone.ID]
+	opt := dnsService.NewListAllDnsRecordsOptions()
+	opt.SetType(string(record.Spec.RecordType))
+	opt.SetName(record.Spec.DNSName)
+	for _, target := range record.Spec.Targets {
+		opt.SetContent(target)
+		result, response, err := dnsService.ListAllDnsRecords(opt)
+		if err != nil {
+			if response.StatusCode != http.StatusNotFound {
+				return fmt.Errorf("delete: failed to list the dns record: %v", err)
+			}
+			continue
+		}
+		if result != nil && result.Result != nil {
+			for _, resultData := range result.Result {
+				if resultData.ID != nil {
```

Comment:
Might be cleaner to check if `resultData.ID == nil` here and return the error on line 121 if it is (rather than indent the rest of the code path)

Reply:
This suggestion might apply to other place as well